include/ultra64/rcp.h

```diff
@@ -1,139 +1,802 @@
+// 8-14 are reserved
+
+/**
+ * Signal Processor (SP) Memory
+ */
+#define SP_DMEM_START   0x04000000
+#define SP_DMEM_END     0x04000FFF
+#define SP_IMEM_START   0x04001000
+#define SP_IMEM_END     0x04001FFF
+
+/**
+ * Signal Processor (SP) CP0 Registers
+ */
+
+#define SP_BASE_REG         0x04040000
+
+// SP memory address (R/W): [12] 0=DMEM,1=IMEM, [11:0] DMEM/IMEM address
+#define SP_MEM_ADDR_REG     (SP_BASE_REG + 0x00)
+
+// SP DRAM DMA address (R/W): [23:0] RDRAM address
+#define SP_DRAM_ADDR_REG    (SP_BASE_REG + 0x04)
+
+// SP read DMA length (R/W): [31:20] skip, [19:12] count, [11:0] length; RDRAM -> I/DMEM
+#define SP_RD_LEN_REG       (SP_BASE_REG + 0x08)
+
+// SP write DMA length (R/W): [31:20] skip, [19:12] count, [11:0] length; I/DMEM -> RDRAM
+#define SP_WR_LEN_REG       (SP_BASE_REG + 0x0C)
+
+// SP status (R/W): [14:0] valid bits; see below for write/read mode
+#define SP_STATUS_REG       (SP_BASE_REG + 0x10)
+
+// SP DMA full (R): [0] dma full
+#define SP_DMA_FULL_REG     (SP_BASE_REG + 0x14)
+
+// SP DMA busy (R): [0] dma busy
+#define SP_DMA_BUSY_REG     (SP_BASE_REG + 0x18)
+
+// SP semaphore (R/W): Read: [0] acquire semaphore; Write: [] release semaphore
+#define SP_SEMAPHORE_REG    (SP_BASE_REG + 0x1C)
+
+// SP PC (R/W): [11:0] program counter
+#define SP_PC_REG       0x04080000
+
+/**
+ * SP_MEM_ADDR_REG: bit 12
+ */
+#define SP_DMA_DMEM     (0 << 12)
+#define SP_DMA_IMEM     (1 << 12)
+
+/**
+ * SP_STATUS_REG: write bits
+ */
+#define SP_CLR_HALT         (1 <<  0)   // clear halt
+#define SP_SET_HALT         (1 <<  1)   //   set halt
+#define SP_CLR_BROKE        (1 <<  2)   // clear broke
+#define SP_CLR_INTR         (1 <<  3)   // clear interrupt
+#define SP_SET_INTR         (1 <<  4)   //   set interrupt
+#define SP_CLR_SSTEP        (1 <<  5)   // clear sstep
+#define SP_SET_SSTEP        (1 <<  6)   //   set sstep
+#define SP_CLR_INTR_BREAK   (1 <<  7)   // clear interrupt on break
+#define SP_SET_INTR_BREAK   (1 <<  8)   //   set interrupt on break
+#define SP_CLR_SIG0         (1 <<  9)   // clear signal 0
+#define SP_SET_SIG0         (1 << 10)   //   set signal 0
+#define SP_CLR_SIG1         (1 << 11)   // clear signal 1
+#define SP_SET_SIG1         (1 << 12)   //   set signal 1
+#define SP_CLR_SIG2         (1 << 13)   // clear signal 2
+#define SP_SET_SIG2         (1 << 14)   //   set signal 2
+#define SP_CLR_SIG3         (1 << 15)   // clear signal 3
+#define SP_SET_SIG3         (1 << 16)   //   set signal 3
+#define SP_CLR_SIG4         (1 << 17)   // clear signal 4
+#define SP_SET_SIG4         (1 << 18)   //   set signal 4
+#define SP_CLR_SIG5         (1 << 19)   // clear signal 5
+#define SP_SET_SIG5         (1 << 20)   //   set signal 5
+#define SP_CLR_SIG6         (1 << 21)   // clear signal 6
+#define SP_SET_SIG6         (1 << 22)   //   set signal 6
+#define SP_CLR_SIG7         (1 << 23)   // clear signal 7
+#define SP_SET_SIG7         (1 << 24)   //   set signal 7
+
+/*
```

Comment:
There's inconsistency in `/**` vs `/*` in here, is it meant to be like that?

Reply:
Since the header is rather large I've tried to split the different components using `/**` comments, so if you ctrl+f for `/**` you can see the different sections, and use `/*` only for particularly long register comments or read/write bits